.github/main.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/dagger/dagger/.github/internal/dagger"
+)
+
+const (
+	daggerVersion      = "v0.12.6"
+	upstreamRepository = "dagger/dagger"
+	defaultRunner      = "ubuntu-latest"
+)
+
+type CI struct {
+	// +private
+	Gha *dagger.Gha
+}
+
+func New(
+	// The dagger repository
+	// +optional
+	// +defaultPath="/"
+	// +ignore=["!.github"]
+	repository *dagger.Directory,
+) *CI {
+	ci := new(CI)
+	ci.Gha = dag.Gha(dagger.GhaOpts{
+		DaggerVersion: daggerVersion,
+		PublicToken:   "dag_dagger_sBIv6DsjNerWvTqt2bSFeigBUqWxp9bhh3ONSSgeFnw",
+		Runner:        ci.BronzeRunner(false),
+		Repository:    repository,
+	})
+	return ci.
+		WithPipeline("Docs", "docs lint", "", false).
+		WithSdkPipelines("python").
```

Comment:
Thinking a bit how this can be translated to the idea of self contained functions in the context of pocketci.

Each of the pipelines described here can be treated as its own self contained function that defines when it should be triggered and where it should run. At the moment we are separating this even further (although I know you already think its overkill) since each step of SDK jobs (lint/test/publish) is its own pipeline with a separate dagger call. If we could bring everything into one we could take an approach that would yield numerous dagger functions (this is still experimental):
```go
// Function only gets called when there is a pull request (opened,reopened,synchronize,ready_for_review) or a push event (to branch main).
func (dev *DaggerDev) SdkGoOnGithub(ctx context.Context,
	src *dagger.Directory,
	// +default="pull_request,push"
	event string,
	// +default="opened,reopened,synchronize,ready_for_review,main"
	filter string,
	eventTrigger *dagger.File) error {
	sdk := dev.SDK().Go

	if err := sdk.Lint(ctx); err != nil {
		return err
	}

	return sdk.Test(ctx)
}
```

Each SDK would have its own function with virtually with the same trigger. This is definitely more verbose than what we currently have (and we still have not solved the compute problem). I'm also not liking so much how the filtering of events and actions within an event is being built, there is some more thinking I need to do there. Its worth doing a bit more thinking and playing with dagger/dagger more to see how quickly the idea of functions defining its own trigger logic breaks.

Reply:
This won't work IMO. You need to keep the CI-specific logic (event routing etc) decoupled, at a higher layer.